BundleTransport/app/src/main/java/net/discdd/bundletransport/TransportToBundleServerManager.java

```diff
@@ -159,34 +163,39 @@ public void onCompleted() {
                     completion.get(Constants.GRPC_LONG_TIMEOUT_MS, TimeUnit.MILLISECONDS);
                 } catch (IOException | ExecutionException | InterruptedException | TimeoutException e) {
                     logger.log(SEVERE, "Failed to download file: " + path, e);
+
                 }
-            }
-        } catch (Exception e) {
-            logger.log(SEVERE, "error: " + e.getMessage());
-            connectError.apply(e);
-        }
 
-        var recencyBlobReq = GetRecencyBlobRequest.newBuilder().setSender(transportSenderId).build();
+                var recencyBlobReq = GetRecencyBlobRequest.newBuilder().setSender(transportSenderId).build();
 
-        var recencyBlob = blockingExchangeStub.withDeadlineAfter(Constants.GRPC_SHORT_TIMEOUT_MS, TimeUnit.MILLISECONDS)
-                .getRecencyBlob(recencyBlobReq);
+                var recencyBlob = blockingExchangeStub.withDeadlineAfter(Constants.GRPC_SHORT_TIMEOUT_MS, TimeUnit.MILLISECONDS)
+                        .getRecencyBlob(recencyBlobReq);
 
-        Path blobPath = fromServerPath.resolve(RECENCY_BLOB_BIN);
-        try (var os = Files.newOutputStream(blobPath, StandardOpenOption.CREATE,
-                                            StandardOpenOption.TRUNCATE_EXISTING)) {
-            logger.log(INFO, "Writing blob to " + blobPath);
-            recencyBlob.writeTo(os);
-        } catch (IOException e) {
-            logger.log(SEVERE, "Failed to write recency blob", e);
-        }
+                Path blobPath = fromServerPath.resolve(RECENCY_BLOB_BIN);
+                try (var os = Files.newOutputStream(blobPath, StandardOpenOption.CREATE,
+                                                    StandardOpenOption.TRUNCATE_EXISTING)) {
+                    logger.log(INFO, "Writing blob to " + blobPath);
+                    recencyBlob.writeTo(os);
+                } catch (IOException e) {
+                    logger.log(SEVERE, "Failed to write recency blob", e);
+                }
 
-        try {
-            channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
-        } catch (InterruptedException e) {
-            logger.log(SEVERE, "could not shutdown channel, error: " + e.getMessage() + ", cause: " + e.getCause());
+                try {
+                    channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
+                } catch (InterruptedException e) {
+                    logger.log(SEVERE, "could not shutdown channel, error: " + e.getMessage() + ", cause: " + e.getCause());
+                }
+                logger.log(INFO, "Connect server completed");
+                connectComplete.apply(null);
+            }
+        } catch (IllegalArgumentException | StatusRuntimeException e) {
+            logger.log(SEVERE, "Failed to connect to server", e);
+            new Handler(Looper.getMainLooper()).post(() -> {
+                Toast.makeText(BundleTransportActivity.applicationContext,
```

Comment:
you should put the application context in the constructor or pass the error back through connectCompete to the fragment create the toast.